src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,157 @@
+#include <limits>
+#include <list>
+
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+/*
+ * This subroutine scans the least desirable available choices until the
+ * accumulated reject_cost() is lower than a required value.
+ * If the thereby accumulated reject_desirability is not lower than a threshold
+ * an iterator to the choice one past the subsequence to reject is returned.
+ * Otherwise the end() iterator is returned, that the constraints are not satisfied
+ * by any continuous subsequence starting at sorted_choices.begin().
+ */
+const std::list<std::shared_ptr<TuningChoice>>::const_iterator sacrifice_choices(
+    const std::list<std::shared_ptr<TuningChoice>>& sorted_choices, float required_cost_delta,
+    float acceptible_desirability_delta = -std::numeric_limits<float>::infinity()) {
+  auto desirability_delta = 0.0f;
+  auto cost_delta = 0.0f;
+
+  auto choice = sorted_choices.cbegin();
+  while (choice != sorted_choices.cend()) {
+    if (cost_delta <= required_cost_delta) {
+      break;
+    }
+    desirability_delta += (*choice)->reject_desirability();
+    cost_delta += (*choice)->reject_cost();
+    ++choice;
+  }
+
+  if (cost_delta <= required_cost_delta && desirability_delta >= acceptible_desirability_delta) {
+    return choice;
+  }
+  return sorted_choices.cend();
+}
+
+std::vector<std::shared_ptr<TuningOperation>> GreedySelector::select(
+    const std::vector<std::shared_ptr<TuningChoice>>& choices, float cost_budget) {
+  std::vector<std::shared_ptr<TuningOperation>> operations;
+  operations.reserve(choices.size());
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
+
+  // Accumulate absolute cost balance from currently chosen choices.
+  float cost_balance = 0.0f;
+  for (const auto& choice : choices) {
+    cost_balance += choice->current_cost();
+  }
+
+  // Desirability balance is always relative to current system state.
+  float desirability_balance = 0.0f;
+
+  // Sort choices:
+  //  1) by reject_cost() ascending
+  //  2) by accept_desirability() ascending stable
+  // The most desirable choice is at back() and among equaly desirable choices
+  // the least costly choices to accept are nearer back().
+  std::list<std::shared_ptr<TuningChoice>> sorted_choices(choices.cbegin(), choices.cend());
+
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->reject_cost() < rhs->reject_cost();
+  });
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->accept_desirability() < rhs->accept_desirability();
+  });
+
+  // If current state exceeds cost_budget,
+  // reject the least desirable choices to reduce cost_balance.
+  if (cost_balance > cost_budget) {
+    LOG_INFO("Cost balance of " << cost_balance << " exceeds budget of " << cost_budget);
+    const auto sacrifice_until = sacrifice_choices(sorted_choices, cost_budget - cost_balance);
```

Comment:
I don‘t get this. Your `sacrifice_choices` method takes a parameter `cost_budget` but you pass in `cost_budget - cost_balance`?
Also, I would consider calling this `last_sacrificed_choice` instead of `sacrifice_until`

Reply:
The parameter that we pass `cost_budget - cost_balance` as is actually called `required_cost_delta` (refer to line 20).

You are right that `sacrifice_until` sounds a little unintuitive but as at the time of the method call no choice has been sacrificed and that `sacrifice_until` ist the first choice that will _not_ be sacrificed (end iterator), we could call it something like `first_choice_not_to_be_sacrificed`, or shorter `sacrifice_choices_until` or `first_salvaged_choice`?

Reply:
1) Did you change something about that? If not, maybe I misread it. This is fine.

2) Since this is a naming decision limited to this place, I don't think we need to discuss this any longer. Let's leave it as it is.